test/Microsoft.NET.Build.Containers.IntegrationTests/TargetsTests.cs

```diff
@@ -10,24 +10,36 @@ namespace Microsoft.NET.Build.Containers.Targets.IntegrationTests;
 [Collection(nameof(MSBuildCollection))]
 public class TargetsTests
 {
-    [InlineData("SelfContained", true, "/app/foo.exe")]
-    [InlineData("SelfContained", false, "dotnet", "/app/foo.dll")]
-    [InlineData("PublishSelfContained", true, "/app/foo.exe")]
-    [InlineData("PublishSelfContained", false, "dotnet", "/app/foo.dll")]
+    [InlineData("win", "SelfContained", true, "C:/app/foo.exe")]
+    [InlineData("win", "SelfContained", false, "dotnet", "C:/app/foo.dll")]
+    [InlineData("win", "PublishSelfContained", true, "C:/app/foo.exe")]
+    [InlineData("win", "PublishSelfContained", false, "dotnet", "C:/app/foo.dll")]
+    [InlineData("linux", "SelfContained", true, "/app/foo")]
+    [InlineData("linux", "SelfContained", false, "dotnet", "/app/foo.dll")]
+    [InlineData("linux", "PublishSelfContained", true, "/app/foo")]
+    [InlineData("linux", "PublishSelfContained", false, "dotnet", "/app/foo.dll")]
```

Comment:
nit: can you make the `rid` strings here valid RIDs (e.g. `win-x64`, `linux-x64`) instead of just the OS portion of a RID? you could also keep them here, call them something other than `rid`, and then on line 31 expand them into a full RID if you like.

I just want to avoid giving a name to a thing that isn't 100% aligned with the name.

Reply:
Yes, I see that. I originally called them `runtimeIdentifiers`, but when I looked at the tests in my test runner, the actual assert disappeared because the parameter list was so long, so I shortened the parameter name. I'd rather not tie it to _one_ arch, as the arch plays no role here. Would a proper name for the "win-" and "linux-" part of the rid be "os" (referring to https://learn.microsoft.com/en-us/dotnet/core/rid-catalog)?

Reply:
Sure, that would make perfect sense in this context 👍

Reply:
Fixed it. `os` is short and readable to:

![image](https://github.com/dotnet/sdk/assets/1628994/baaabca5-e2e5-4da6-b371-e8189f52cbe3)

Reply:
Blah! Looks like the tests are os-specific. This also means that cross-compiling will not work, doesn't it? This would make the path to the executable and DLL if building a Windows-container _on windows_ be different than if building a Windows-container on *nix...

Do you know _where_ in the .targets files the variable for the directory separator character is stored? I need to try to look at a fix for this too. Because when building containers, this should be the path separator character of OS of the target container, not the build host

The helix build (which probably runs on Windows) failed with the following error:

```bash
    Microsoft.NET.Build.Containers.Targets.IntegrationTests.TargetsTests.CanDeferContainerAppCommand(os: "win", prop: "SelfContained", value: True, expectedAppCommandArgs: ["C:/app/foo.exe"]) [FAIL]
      Expected root[0] to be "C:/app/foo.exe", but "C:\app/foo.exe" differs near "\ap" (index 2).
```

Reply:
Actually it might make more sense to put the slash on the ContainerWorkingDirectory property value, instead of putting that knowledge in the creation of the `ContainerAppCommand` values - this would actually be more aligned with other MSBuild Properties that are paths - for example PublishDir, OutputDir, etc.

@rainersigwald this sound reasonable to you?

Reply:
Yes, I see your point. However, _something_ is replacing whichever path separator used in the .targets file with the one on the current build OS, it seems. Because, on my mac, it is **C:/app/foo.exe**, not **C:\app/foo.exe**, when targeting Windows, even though it is set to **C:\app/foo.exe** (computed) in the .targets file. So, something, _somewhere_ is normalising the paths to whatever it is on the _build_ platform. But, it should be normalised to what is used on the _target_ platform instead.

I was thinking, are there any other build targets than containers, that are dependent on slashes being the right type on the _target_ platform? I'm trying to browse through the sdk source code, but this is the first time I'm looking at it, so I might be looking in the wrong places.

A couple of candidates I'm looking at, are:

* src/Cli/dotnet/CommandFactory/CommandResolution/PackagedCommandSpecFactory.cs
* src/Cli/dotnet/CommandFactory/CommandResolution/OutputPathCommandResolver.cs

These both use: src/Cli/Microsoft.DotNet.Cli.Utils/PathUtility.cs

but I might be looking in the wrong places.

Do you know which targets are run when publishing to a docker image? And which one of them could be normalising the paths? I could look at trying to fix it so that they use the target path, but I don't want to mess up anything. I think changing PathUtility.cs is high risk, as it is probably used for everything all and around

Reply:
So, I've looked a bit more into this, @baronfel , and, unfortunately, I've traced the issue all the way down to another repository, dotnet/msbuild. The bug/feature is that when instantiating a `TaskItem`, all path separators are "fixed" (i.e. fixed for the platform you are _building on_). But, when building a docker image, you want the items to have path separators for the platform you are building _for_.

See the TaskItem source here: https://github.com/dotnet/msbuild/blob/863209471a60e9dfd387ed54d4edb85fd8db57ee/src/Utilities/TaskItem.cs#L80

The build task CreateNewImage uses ITaskItem's for all properties, and, consequently, the paths of all of the properties here will be changed to have the path separator of the platform one is building on.

So, then, the question is, what do we do here? I see the question with the path normalisation to be a bit of a bigger, more wide-ranging issue than just my pull request, and I don't feel it's natural to solve it as a part of this PR.

Would you be happy if I:

1. Change the unit test so that it checks that the deferred app command checks that it's the string `$"C:{Path.DirectorySeparatorChar}app{Path.DirectorySeparatorChar}foo.exe` on windows? This means that if you build a windows docker image on *nix, it _will_ have the entrypoint set to `C:/app/foo.exe`, not `C:\app\foo.exe`. I guess this should work, but it's a bit wonky that the build results are dependent on the platform
2. Create a separate issue where we try to create a plan for how to make the `CreateNewImage` task takes the _target_ platform into account when normalising the paths? This could involve creating a different implementation of `ITaskItem` that normalises the paths dependent on the target platform, not the platform you're building on, or it could be something entirely different.

Or would this make things ugly?

As a note, you can see the same problem in the tests for `CreateNewImage`, if you add paths to the appCommand or entryPoint arguments (test/Microsoft.NET.Build.Containers.UnitTests/CreateNewImageTests.cs)

Example of (new, not pushed) failing tests are:

```csharp
    [Theory]
    // Use Windows-style paths. Expect that paths are _not_ converted to Unix-style.
    [InlineData("", "", "C:\\app\\appCommand", "", "", "", null, new[] { "C:\\app\\appCommand" })]
    // Use Unix-style paths. Expect that paths are _not_ converted to Windows-style.
    [InlineData("", "", "/app/appCommand", "", "", "", null, new[] { "/app/appCommand" })]
    public void EntrypointAndCmd_uses_target_platform_path_separators(string entrypoint, string entrypointArgs, string appCommand, string appCommandArgs, string defaultArgs, string? baseImageEntrypoint, string[]? expectedEntrypoint, string[]? expectedCmd)
        => ValidateArgsAndCmd("DefaultArgs", entrypoint, entrypointArgs, appCommand, appCommandArgs, defaultArgs, baseImageEntrypoint, expectedEntrypoint, expectedCmd);
```

They fail for the Windows version when running on my Mac. They might pass on Windows, I'm not sure.

Reply:
I reported a bug in the dotnet/msbuild repo as well. I'm not sure whether it's a bug with the TaskItem, or with the usage of it in the CreateNewImage.

Just FYI. https://github.com/dotnet/msbuild/issues/10121

Reply:
excellent analysis and bug report - I knew that Items did some amount of normalization by default, but hadn't done the kind of exhaustive comparison that you've done here to log the issue/try to systematically address it.

Reply:
Thanks. I'm not sure which tree is the correct one to be barking up, though. Is it a bug? Or are we using it incorrectly? Are there other use cases that might need a different path separator than the build host or is the building of container for a different platform the only one?

I guess someone needs to discuss this a bit internally, to come up with suggestions on what the best solution is, both risk/reward wise, and most forward-looking.

Please feel free to hold off or merge this PR, depending on what works best for you, and do let me know if I can contribute further on the TaskItem usage (or bug)